google/resource_compute_instance_from_template.go

```diff
@@ -0,0 +1,136 @@
+package google
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	strcase "github.com/stoewer/go-strcase"
+)
+
+func resourceComputeInstanceFromTemplate() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceComputeInstanceFromTemplateCreate,
+		Read:   resourceComputeInstanceRead,
+		Update: resourceComputeInstanceUpdate,
+		Delete: resourceComputeInstanceDelete,
+
+		// Import doesn't really make sense, because you could just import
+		// as a google_compute_instance.
+
+		Timeouts: resourceComputeInstance().Timeouts,
+
+		Schema:        computeInstanceFromTemplateSchema(),
+		CustomizeDiff: resourceComputeInstance().CustomizeDiff,
+	}
+}
+
+func computeInstanceFromTemplateSchema() map[string]*schema.Schema {
+	s := resourceComputeInstance().Schema
+
+	for _, field := range []string{"boot_disk", "machine_type", "network_interface"} {
+		s[field].Required = false
+	}
+
+	recurseOnSchema(s, func(field *schema.Schema) {
+		// We don't want to accidentally use default values to override the instance
+		// template, so remove defaults.
+		field.Default = nil
+
+		// Make non-required fields computed since they'll be set by the template.
+		// Leave deprecated and removed fields alone because we don't set them.
+		if !field.Required && !(field.Deprecated != "" || field.Removed != "") {
+			field.Computed = true
+		}
+	})
+
+	s["source_instance_template"] = &schema.Schema{
+		Type:     schema.TypeString,
+		Required: true,
+		ForceNew: true,
+	}
+
+	return s
+}
+
+func recurseOnSchema(s map[string]*schema.Schema, f func(*schema.Schema)) {
+	for _, field := range s {
+		f(field)
+		if e := field.Elem; e != nil {
+			if r, ok := e.(*schema.Resource); ok {
+				recurseOnSchema(r.Schema, f)
+			}
+		}
+	}
+}
+
+func resourceComputeInstanceFromTemplateCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	// Get the zone
+	z, err := getZone(d, config)
+	if err != nil {
+		return err
+	}
+	log.Printf("[DEBUG] Loading zone: %s", z)
+	zone, err := config.clientCompute.Zones.Get(project, z).Do()
+	if err != nil {
+		return fmt.Errorf("Error loading zone '%s': %s", z, err)
+	}
+
+	instance, err := computeInstance(project, zone, d, config)
+	if err != nil {
+		return err
+	}
+
+	// Force send all top-level fields in case they're overridden to zero values.
+	// TODO: consider doing so for nested fields as well.
+	for f, s := range computeInstanceFromTemplateSchema() {
+		// It seems that GetOkExists always returns true for sets.
+		// TODO: confirm this and file issue against Terraform core.
+		// In the meantime, don't force send sets.
+		if s.Type == schema.TypeSet {
+			continue
+		}
+
+		if _, exists := d.GetOkExists(f); exists {
+			instance.ForceSendFields = append(instance.ForceSendFields, strcase.UpperCamelCase(f))
```

Comment:
:/ I can see this breaking when someone adds a field to `compute_instance` that isn't a direct 1:1 snake_case version of the API field.

Reply:
Fair. I took a look at our current fields, and as far as I can tell, things that would potentially be overridden to their zero value are indeed snake_case versions of the API fields. Given that we're generally trying to do that sort of thing anyway for new fields, I'm not too concerned, but I added a comment.

Reply:
I assume the API returns a good error if a field that doesn't match a field is set in ForceSendFields? I'm gonna test that real quick. As long as a field that doesn't match up isn't silently ignored, I think this is a manageable solution. Nested values scare me a little, and I think over time we may want to add a test that, e.g., checks all the properties work out to the same strings as ForceSendFields, but that's probably not a project for today.

Reply:
Nope, it's silently ignored. I could instead directly set the fields that we expect to be able to be overridden to zero values, but I honestly think that's just about as error prone as this is.

Reply:
The API does, actually, just silently ignore invalid fields. :/ I've opened google/google-api-go-client#280 to see if I can't get the client to throw an error on invalid fields. Otherwise, this may be something worth bringing in internally, but I suppose I'm happy to punt on that, as we have test coverage, and don't have perfect `ForceSendFields` support _elsewhere_, anyways, so the class of bugs this would enable is a class of bugs we already accept.

Sorry, this is basically a long comment explaining for future-Paddy/future maintainers why this was an acceptable risk, in case someone's doing archaeology for a bug. :)